libraries/core_libs/consensus/src/pbft/pbft_manager.cpp

```diff
@@ -1110,14 +1123,50 @@ blk_hash_t PbftManager::generatePbftBlock(const blk_hash_t &prev_blk_hash, const
   return pbft_block->getBlockHash();
 }
 
-std::shared_ptr<Vote> PbftManager::generateVote(blk_hash_t const &blockhash, PbftVoteTypes type, uint64_t round,
-                                                size_t step) {
+std::shared_ptr<Vote> PbftManager::generateVote(blk_hash_t const &blockhash, PbftVoteTypes type, uint64_t period,
+                                                uint64_t round, size_t step) {
   // sortition proof
-  VrfPbftSortition vrf_sortition(vrf_sk_, {type, round, step});
+  VrfPbftSortition vrf_sortition(vrf_sk_, {type, period, round, step});
   return std::make_shared<Vote>(node_sk_, std::move(vrf_sortition), blockhash);
 }
 
-uint64_t PbftManager::getThreshold(PbftVoteTypes vote_type) const {
+std::pair<bool, std::string> PbftManager::validateVote(const std::shared_ptr<Vote> &vote) const {
+  const uint64_t vote_period = vote->getPeriod();
+
+  // Validate vote against dpos contract
+  try {
+    const auto pk = key_manager_->get(vote->getVoterAddr());
+    if (!pk) {
+      std::stringstream err;
+      err << "No vrf key mapped for vote author " << vote->getVoterAddr();
+      return {false, err.str()};
+    }
+
+    // TODO[1896]: final_chain_->dpos... ret values should be cached as it is heavily used and most of the time it
+    // returns the same value!
+    const uint64_t voter_dpos_votes_count = final_chain_->dpos_eligible_vote_count(vote_period, vote->getVoterAddr());
+    const uint64_t total_dpos_votes_count = final_chain_->dpos_eligible_total_vote_count(vote_period);
+    const uint64_t pbft_sortition_threshold = getPbftSortitionThreshold(vote->getType(), vote_period);
+
```

Comment:
I think this should be called only when `vote_period != dpos_period_` as we currently have some kind of caching we should use it

Reply:
https://github.com/Taraxa-project/taraxa-node/issues/1896